### PR TITLE
UI updates 

### PR DESCRIPTION
**Pull request by levitte**
##### Checklist

- [x] documentation is added or updated
- [x] CLA is signed

##### Description of change

- constify the input parameter to UI method getters.  *This is ABI compatible, it merely makes an already existing practice explicit*
- add space for an application data field in the UI structure, and functionality to access it.
- add a UI_METHOD wrapper for the PEM password callback, for applications that need to pass the latter to an interface that takes a UI_METHOD.
- document UI utility functions

### CODE REVIEW

**Review by richsalz — CHANGES_REQUESTED** [review]
any reason why not to use the standard EXDATA ?

**Comment by levitte** [comment]
> any reason why not to use the standard EXDATA ?

not really, just that it escaped my mind.  Will fix

File: doc/man3/UI_UTIL_read_pw.pod [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                     int verify);
+ UI_METHOD *UI_UTIL_wrap_read_pem_callback(pem_password_cb *cb, int
+ rwflag);
+
+=head1 DESCRIPTION
+
+UI_UTIL_read_pw_string() asks for a passphrase, using B<prompt> as a
+prompt, and stores it in B<buf>.
+The maximum allowed size is given with B<length>, including the ending
+NUL byte.
```
Comment (kaduk) [inline doc/man3/UI_UTIL_read_pw.pod]: Maybe "trailing NUL", "NUL terminator", or "terminating NUL byte"?  No real preference from me, though.

**Comment by kroeckx** [comment]
Please propose some policy about which changes are allowed in public headers.

**Comment by kroeckx** [comment]
Is there a reason this isn't tested?

**Comment by levitte** [comment]
> Please propose some policy about which changes are allowed in public headers.

Do these suffice? They say nothing about headers and everything about backward ABI compatibility. 

https://www.openssl.org/docs/faq.html#MISC8
https://www.openssl.org/policies/releasestrat.html

The assumption is that as long as the API doesn't break applications and ABI compatibility is maintained, we should be fine.

**Comment by levitte** [comment]
> Is there a reason this isn't tested?

I'm writing a test, will show up in a commit in a couple of hours.

File: doc/man3/UI_UTIL_read_pw.pod (hunk 1 above)
Comment (levitte) [inline doc/man3/UI_UTIL_read_pw.pod]: Thanks. Will change

**Comment by levitte** [comment]
I've made a number of updates:

- changed the UI_METHOD app data to be an ex_data field (ping @richsalz)
- added a test program that tests the UI_METHOD wrapper as well as the method already tested through the `openssl` command.  This simplified the UI test recipe.

**Review by richsalz — APPROVED** [review]
Looks okay to me.

**Comment by levitte** [comment]
Pushed a couple of fixes for the travis build issues

**Comment by levitte** [comment]
Builds are passing now.  @richsalz, would you mind reconfirming?

**Comment by richsalz** [comment]
sure, +1

**Comment by levitte** [comment]
Perfect.  Thanks.  Merged

**Pull request closed without merging**